File: android/app/src/main/java/com/myvu/client/reminder/BootReceiver.java
package com.myvu.client.reminder;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import com.myvu.client.core.LogBus;
import com.myvu.client.database.Reminder;
import com.myvu.client.database.ReminderRepository;

import java.util.List;

public class BootReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        if (intent == null) return;
        String action = intent.getAction();
        if (Intent.ACTION_BOOT_COMPLETED.equals(action) ||
            Intent.ACTION_TIME_CHANGED.equals(action) ||
            Intent.ACTION_TIMEZONE_CHANGED.equals(action)) {

            LogBus.log("BootReceiver -> Rescheduling pending reminders after " + action);

            ReminderRepository repo = new ReminderRepository(context);
            List<Reminder> pending = repo.getPendingReminders();

            long now = System.currentTimeMillis();
            for (Reminder r : pending) {
                if (r.getTriggerAt() > now) {
                    ReminderScheduler.scheduleReminder(context, r.getId(), r.getTriggerAt(), r.getAlarmRequestCode());
                } else {
                    // Mark past un-triggered reminders as FAILED or trigger immediately
                    repo.updateReminderState(r.getId(), "FAILED");
                }
            }
        }
    }
}

File: android-kotlin/app/src/main/java/com/myvu/client/reminder/BootReceiver.kt
package com.myvu.client.reminder

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.myvu.client.core.LogBus
import com.myvu.client.core.ServiceKeepAliveHelper
import com.myvu.client.database.ReminderRepository

class BootReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context?, intent: Intent?) {
        if (context == null || intent == null) return
        val action = intent.action
        LogBus.log("BootReceiver -> Received broadcast action: $action")

        if (Intent.ACTION_BOOT_COMPLETED == action ||
            Intent.ACTION_MY_PACKAGE_REPLACED == action ||
            Intent.ACTION_POWER_CONNECTED == action ||
            Intent.ACTION_USER_PRESENT == action ||
            Intent.ACTION_TIME_CHANGED == action ||
            Intent.ACTION_TIMEZONE_CHANGED == action
        ) {
            try {
                ServiceKeepAliveHelper.ensureServiceRunning(context)
                LogBus.log("BootReceiver -> Ensured MyvuService running on $action")
            } catch (e: Exception) {
                LogBus.error("BootReceiver -> Failed to ensure MyvuService running", e)
            }

            if (Intent.ACTION_BOOT_COMPLETED == action ||
                Intent.ACTION_TIME_CHANGED == action ||
                Intent.ACTION_TIMEZONE_CHANGED == action
            ) {
                LogBus.log("BootReceiver -> Rescheduling pending reminders after $action")

                val repo = ReminderRepository(context)
                val pending = repo.getPendingReminders()

                val now = System.currentTimeMillis()
                for (r in pending) {
                    if (r.triggerAt > now) {
                        ReminderScheduler.scheduleReminder(context, r.id, r.triggerAt, r.alarmRequestCode)
                    } else {
                        repo.updateReminderState(r.id, "FAILED")
                    }
                }
            }
        }
    }
}
